### PR TITLE
support repetition ngram logits processor

### PR DESCRIPTION
**Pull request by grimoire**
repetition stopping is implemented as logits processor.
If we turn it to an engine-level feature, the implementation would be much easy and the performance would be better.

### CODE REVIEW

File: lmdeploy/pytorch/engine/logits_process.py [hunk 1]
```diff
@@ -68,6 +70,142 @@ def _filter_minp_sorted_(scores: torch.Tensor, minp: torch.Tensor, filter_value:
...
+    """
+
+    batch_size, seq_len = token_ids.size()
+    if seq_len < max_n:
+        # Not enough tokens to form a single n-gram
+        matched_mask = torch.zeros((batch_size, 0), dtype=torch.bool, device=token_ids.device)
+        found = torch.zeros((batch_size, ), dtype=torch.bool, device=token_ids.device)
+        return matched_mask, found
+    # token_ids could be 0, so we add 1 to avoid div 0
+    token_ids = token_ids.to(torch.float32).log2() + 1
```
Comment (lvhan028) [inline lmdeploy/pytorch/engine/logits_process.py]: It is supposed to be `token_ids = (token_ids + 1).to(torch.float32).log2()`, isn't it?

File: lmdeploy/messages.py [hunk 2]
```diff
@@ -129,6 +129,10 @@ class GenerationConfig:
     # router replay
     return_routed_experts: bool = False
 
+    # ngram
+    ngram_size: int = 0
+    ngram_threshold: int = 0
+
```
[the remark this answers is not in the document]
Comment (lvhan028) [reply lmdeploy/messages.py:137]: agreed

File: lmdeploy/messages.py [hunk 3]
```diff
@@ -129,6 +129,11 @@ class GenerationConfig:
     # router replay
     return_routed_experts: bool = False
 
+    # ngram
+    repetition_ngram_size: int = 0
+    repetition_ngram_threshold: int = 0
+    repetition_ngram_window_size: int = 1024
+
```
Comment (lvhan028) [inline lmdeploy/messages.py:137]: I suggest keeping `repetition_ngram_window_size` as an internal constant rather than exposing it in the public API. A fixed value of 1024 should be sufficient for most use cases

File: lmdeploy/pytorch/engine/logits_process.py [hunk 4]
```diff
@@ -68,6 +70,142 @@ def _filter_minp_sorted_(scores: torch.Tensor, minp: torch.Tensor, filter_value:
...
+
+    same_window = window_size is None
+    if not same_window:
+        # fill -1 for window_size < max_window_size
+        mask = torch.arange(max_window_size,
+                            device=token_ids.device).unsqueeze(0) >= (max_window_size - window_size.unsqueeze(1))
+        token_ids.masked_fill_(~mask, -1)
+
+    # normalize ids
+    same_n = n is None
```
Comment (lvhan028) [inline lmdeploy/pytorch/engine/logits_process.py:161]: Relying on `n is None` to indicate that all sequences share the same `n` is implicit and easy to overlook, especially without closely reading `sampling.py`.
To improve clarity, could we pass the explicit n value to this function instead?

File: lmdeploy/pytorch/engine/logits_process.py [hunk 5]
```diff
@@ -68,6 +70,142 @@ def _filter_minp_sorted_(scores: torch.Tensor, minp: torch.Tensor, filter_value:
...
+    ones_ids = torch.ones_like(normed_n_ids)
+    if not same_n:
+        # fill 0 for n < max_n
+        normed_n_ids = normed_n_ids * mask.to(torch.float32)
+        ones_ids = ones_ids * mask.to(torch.float32)
+    normed_ids = torch.cat([normed_ids, normed_ids_p2], dim=0)
+    normed_n_ids = torch.cat([normed_n_ids, ones_ids], dim=0)
+
+    # check cos distance & check vector length
+    match_norm = torch.conv1d(normed_ids.unsqueeze(0), normed_n_ids.unsqueeze(1), groups=batch_size * 2)[0]
```
Comment (lvhan028) [inline lmdeploy/pytorch/engine/logits_process.py:182]: brilliant!

File: lmdeploy/messages.py (hunk 3 above)
Comment (lvhan028) [reply lmdeploy/messages.py:137]: Based on this PR, my understanding is that for any sequence configured with n-gram constraints, the engine checks for duplicates within the sliding window of the last `max_ngram_window_size` tokens (i.e., `token_ids[-max_ngram_window_size:]`). Once the repetition frequency exceeds the threshold, the engine prevents further token generation. Please correct me if I'm mistaken.
Since this behavior heavily depends on the `max_ngram_window_size` setting, I'm considering an alternative approach: perform the n-gram check only every `max_ngram_window_size` steps. Specifically, we could trigger the check when the length of generated_token_ids is divisible by `max_ngram_window_size` (i.e., len % size == 0). What do you think?"

Comment (grimoire) [reply lmdeploy/messages.py:137]: > perform the n-gram check only every max_ngram_window_size steps. Specifically, we could trigger the check when the length of generated_token_ids is divisible by max_ngram_window_size (i.e., len % size == 0).

Hard to do this in batch.

File: lmdeploy/messages.py [hunk 6]
```diff
@@ -129,6 +129,11 @@ class GenerationConfig:
     # router replay
     return_routed_experts: bool = False
 
+    # ngram
+    repetition_ngram_size: int = 0
+    repetition_ngram_threshold: int = 0
```
Comment (RunningLeon) [inline lmdeploy/messages.py:136]: Most users don't know how to set them.  Better add docstring for these two arguments.

File: lmdeploy/messages.py [hunk 7]
```diff
@@ -129,6 +131,10 @@ class GenerationConfig:
     # router replay
     return_routed_experts: bool = False
 
+    # ngram, generation would stop if latest [size] tokens are repeated for [threshold] times
+    repetition_ngram_size: int = 0
```
Comment (RunningLeon) [inline lmdeploy/messages.py:135]: how do we use this feature for api_server?

Comment (grimoire) [reply lmdeploy/messages.py:135]: I don't know how to add sampling args
